test/integration/scheduler/local-pv-neg-affinity_test.go

```diff
@@ -0,0 +1,313 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+// This file tests the VolumeScheduling feature.
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/api/testapi"
+	"k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	"k8s.io/kubernetes/pkg/controller/volume/persistentvolume"
+	"k8s.io/kubernetes/plugin/pkg/scheduler"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/factory"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	affinityLabelKey = "kubernetes.io/hostname"
+)
+
+func TestLocalPVNegativeAffinity(t *testing.T) {
+	config := setupNodes(t, "volume-scheduling", 3)
+	defer config.teardown()
+
+	pv := makeHostBoundPV(t, "local-pv", classImmediate, "", "", "node-1")
+	pvc := makePVC("local-pvc", config.ns, &classImmediate, "")
+
+	// Create PV
+	if _, err := config.client.CoreV1().PersistentVolumes().Create(pv); err != nil {
+		t.Fatalf("Failed to create PersistentVolume %q: %v", pv.Name, err)
+	}
+
+	// Create PVC
+	if _, err := config.client.CoreV1().PersistentVolumeClaims(config.ns).Create(pvc); err != nil {
+		t.Fatalf("Failed to create PersistentVolumeClaim %q: %v", pvc.Name, err)
+	}
+
+	nodeMarkers := []interface{}{
+		markNodeAffinity,
+		markNodeSelector,
+		markNodeName,
+	}
+
+	for i := 0; i < len(nodeMarkers); i++ {
+		pod := makePod("local-pod", config.ns, []string{"local-pvc"})
+		nodeMarkers[i].(func(*v1.Pod, string))(pod, "node-2")
+		// Create Pod
+		if _, err := config.client.CoreV1().Pods(config.ns).Create(pod); err != nil {
+			t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+		}
+
+		if err := waitForPodToSchedule(config.client, pod); err != nil {
```

Comment:
Can you validate that the pod failed to schedule?

It would be good if we could validate the Pod events and look for the predicate filters.  You may need to add some logic in the setup to start the scheduler with an event recorder.

Reply:
done